lib/rate-limit.js

```diff
@@ -1,86 +1,44 @@
- */
-export default function rateLimit (fn, concurrency, delay) {
-  concurrency = positiveInteger('concurrency', concurrency)
-  delay = positiveInteger('delay', delay)
-  const callQueue = []
-  let inFlight = 0
-
-  function shift () {
-    if (inFlight >= concurrency) {
-      return
+export default function rateLimit (instance, maxRetry = 10) {
+  instance.interceptors.response.use(function (response) {
+    // we don't need to do anything here
+    return response
+  }, function (error) {
+    const {config} = error
+    if (!config) {
+      return Promise.reject(error)
     }
-
-    const start = new Date().getTime()
-
-    if (callQueue.length) {
-      const call = callQueue.shift()
-      inFlight++
-
-      let result
-      try {
-        // call the wrapped method
-        const tmp = Promise.resolve(fn.apply(call.self, call.args))
-        // resolve the call with a possibly rejected promise
-        call.resolve(tmp)
-        // rate-limiting doesn't care about errors
-        result = tmp.catch(() => {})
-      } catch (err) {
-        call.reject(err)
-        result = Promise.resolve()
+    config.attemp = config.attemp || 0
+    if (error.response.status >= 500) {
+      config.attemp++
+      config.lastTimeSinceError = config.lastTimeSinceError || Date.now()
+      if (Date.now() - config.lastTimeSinceError >= 5000) {
+        // we reset the attemp here if there is a gap of 5 sec
+        config.attemp = 0
+      }
+      // we reject if there are too much errors in a short period of time
+      if (config.attemp >= maxRetry) {
+        return Promise.reject(error)
       }
-
-      // If this request completed faster than our rate-limit would
-      // allow, wait the appropriate amount of time before marking the
-      // request as finished.
-      result.then(maybeWait).then(goToNextCall)
     }
 
-    function maybeWait () {
-      const duration = start - new Date().getTime()
-      if (duration < delay) {
-        return promisedWait(delay - duration)
+    // retry if we receive 429 or 500
+    if (error.response.status === 429 || error.response.status >= 500) {
+      var rateLimitHeaderValue = 0
+      if (error.response.headers && error.response.headers['X-Contentful-RateLimit-Reset']) {
+        rateLimitHeaderValue = error.response.headers['X-Contentful-RateLimit-Reset']
       }
-    }
-  }
 
-  function goToNextCall () {
-    inFlight--
-    shift()
-  }
+      var wait = rateLimitHeaderValue || Math.pow(2, config.attemp)
+      // convert to ms and add jitter
+      wait *= Math.floor(1000 * ((Math.random() * 1.2) + 1))
 
-  return function (/* wrapped method arguments */) {
-    const self = this
-    const args = Array.prototype.slice.call(arguments)
-    return new Promise((resolve, reject) => {
-      callQueue.push({
-        reject: reject,
-        resolve: resolve,
-        self: self,
-        args: args
+      console.info(`error occured waiting for ${wait}ms before retrying....`)
```

Comment:
I would add at leas a `.` (dot) after the `error ocurred` chunk. Otherwise it could be read as if the error happened while waiting.